### PR TITLE
Improve monitoring class

### PR DESCRIPTION
**Pull request by sebasfalcone**
|Related issue|
|---|
| #25416 |

## Description
This PR improves upon the monitoring to be more reliable

To do so:
- Initializes all servers with its corresponding status
- Updates the http-request submodule to make use of timeouts
- Improves the headers used to get the indexer information to one easier-to-parse

### CODE REVIEW

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 1]
```diff
@@ -86,78 +86,91 @@ class Monitoring final
...
     {
-        // Initialize the map with the values, all servers are available.
-        for (auto& value : values)
+        /**
+         * @brief Health check function
+         *
+         * @note It sends a request to the \p serverAddress and update the \p serverStatus (true if the server is
+         * available, false otherwise).
+         */
+        const auto healthCheck = [this, secureCommunication](const std::string& serverAddress, bool& serverStatus)
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:106]:
```suggestion
        const auto healthCheck = [this](const std::string& serverAddress, bool& serverStatus, const auto& secureCommunication)
```

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 2]
```diff
@@ -167,15 +180,15 @@ class Monitoring final
...
      * @return true if available.
      * @return false if not available.
      */
-    bool isAvailable(const std::string& value)
+    bool isAvailable(const std::string& serverAddress)
     {
         // Check if the server is available.
         std::lock_guard<std::mutex> lock(m_mutex);
-        return m_values.at(value);
+        return m_servers.at(serverAddress);
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:191]:
```suggestion
        std::lock_guard lock(m_mutex);
```

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 3]
```diff
@@ -86,78 +86,91 @@ class Monitoring final
...
-                                    {
-                                        value.second = false; // LCOV_EXCL_LINE
-                                    }
-                                },
-                                [&](const std::string& error, const long /*statusCode*/) { value.second = false; },
-                                "",
-                                headers,
-                                secureCommunication);
+                            healthCheck(serverAddress, serverStatus);
                         }
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:174]:
```suggestion
                            healthCheck(serverAddress, serverStatus, secureCommunication);
```

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 4]
```diff
@@ -86,78 +86,91 @@ class Monitoring final
...
-            [this, secureCommunication]()
+            [this, secureCommunication, healthCheck]()
             {
-                const std::unordered_set<std::string> headers {"Accept-Charset: utf-8"};
                 while (!m_stop)
                 {
-                    std::unique_lock<std::mutex> lock(m_mutex);
                     // Wait for the interval.
+                    std::unique_lock<std::mutex> lock(m_mutex);
                     m_condition.wait_for(lock, std::chrono::seconds(m_interval), [this]() { return m_stop.load(); });
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:165]:
```suggestion
                    std::unique_lock lock(m_mutex);
```

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 5]
```diff
@@ -86,78 +86,91 @@ class Monitoring final
...
+                HttpURL(serverAddress + "/_cat/health?v"),
+                [&](std::string response)
+                {
+                    // Remove the tabs and double spaces.
+                    Utils::replaceAll(response, "\t", " ");
+                    response = Utils::trimRepeated(response, ' ');
+                    // Split the response by rows.
+                    const auto rows {Utils::split(response, '\n')};
+
+                    // Check if the response has the expected number of rows.
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:119]:
```suggestion
                    serverStatus = false;
```

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 6]
```diff
@@ -86,78 +86,91 @@ class Monitoring final
...
+                    else
+                    {
+                        serverStatus = false; // LCOV_EXCL_LINE
+                    }
+                },
+                [&](const std::string& error, const long /*statusCode*/) { serverStatus = false; },
+                "",
+                {"Accept-Charset: utf-8"},
+                secureCommunication);
+        };
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:146]:
```suggestion
                authentication);
```

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 7]
```diff
@@ -86,78 +86,91 @@ class Monitoring final
...
+                            serverStatus = false;
+                        }
+                    }
+                    else
+                    {
+                        serverStatus = false; // LCOV_EXCL_LINE
+                    }
+                },
+                [&](const std::string& error, const long /*statusCode*/) { serverStatus = false; },
+                "",
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:143]:
```suggestion
                [&serverStatus](const std::string& /*error*/, const long /*statusCode*/) { serverStatus = false; },
```

File: src/shared_modules/indexer_connector/src/monitoring.hpp (hunk 1 above)
Comment (Dwordcito) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:106]: this std::function could be moved to a private method.

**Review by Dwordcito — CHANGES_REQUESTED** [review]
HTTP request calls have specified a timeout if some calls are slowly? Take care with this, because is inner the init thread.

File: src/shared_modules/indexer_connector/src/monitoring.hpp (hunk 5 above)
Comment (sebasfalcone) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:119]: I don't fully undertand this request

**Comment by sebasfalcone** [comment]
> HTTP request calls have specified a timeout if some calls are slowly? Take care with this, because is inner the init thread.

This is still pending, I will investigate it tomorrow

File: src/shared_modules/indexer_connector/src/monitoring.hpp (hunk 5 above)
Comment (Dwordcito) [reply src/shared_modules/indexer_connector/src/monitoring.hpp]: this assignation is to avoid the else statement and reduce the complexity/length of this method/std::function

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 8]
```diff
@@ -90,87 +168,32 @@ class Monitoring final
...
-            {
-                healthCheck(serverAddress, m_servers[serverAddress]);
-            }
-        }
+        initialize(serverAddresses, secureCommunication);
 
         // Start the thread, that will check the health of the servers.
         m_thread = std::thread(
-            [this, secureCommunication, healthCheck]()
+            [this, &secureCommunication]()
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:182]: warning with this, maintain secureCommunication as copy, because the lifetime of this variable is only guaranted by the IndexerConnector constructor.

IndexerConnector() local variable -> ServerSelector() pass by reference -> Monitoring() pass by reference -> std::thread pass by reference.

When the IndexerConnector constructor returns, this variable is destroyed.

**Review by Dwordcito — CHANGES_REQUESTED** [review]

**Comment by sebasfalcone** [comment]
## Update
I've changed the way we initialize the server to prevent locking the main thread

As we can see, a slow server will impact the monitoring class:
```
[ RUN      ] MonitoringTest.TestSlowServer
[       OK ] MonitoringTest.TestSlowServer (104 ms)
```

But the initialization is performed inside the working thread so it won't block the main thread:
https://github.com/wazuh/wazuh/blob/7c3e8702aa7b9cee2bcf8c7626fc5bb76b51aca2/src/shared_modules/indexer_connector/src/monitoring.hpp#L178-L184

Also, the mutex is released after verifying each server during initialization:
https://github.com/wazuh/wazuh/blob/7c3e8702aa7b9cee2bcf8c7626fc5bb76b51aca2/src/shared_modules/indexer_connector/src/monitoring.hpp#L139-L152

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 9]
```diff
@@ -86,96 +172,60 @@ class Monitoring final
...
+                            healthCheck(serverAddress, authentication);
                         }
                     }
                 }
             });
+
+        // Wait for the initialization to complete or the timeout to expire.
+        std::unique_lock initLock(m_initMutex);
+        m_initCondition.wait_for(
+            initLock, std::chrono::seconds(INITIALIZATION_TIMEOUT), [this]() { return m_initialized; });
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:215]: It is better to avoid this logic, which is complicated and can also cause certain exceptions if the timeout occurs.


Let's update http request to have the timeout benefits that it has.


https://github.com/wazuh/wazuh-http-request/blob/5807c8beaf5cf15fddb519253acef25b661353c3/src/HTTPRequest.cpp#L46


It is important that we initialize the vector to false for all servers, to avoid an access exception in the isAvailable() method

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/shared_modules/indexer_connector/src/monitoring.hpp (hunk 9 above)
Comment (sebasfalcone) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:215]: Making this change will require a lot of code tweaking, as the interface for http-request changed on this PR
- https://github.com/wazuh/wazuh-http-request/commit/ce95de12171c55aec27393f869ff9035fdd8d63f

Do you want me to proceed?

Comment (Dwordcito) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:215]: yes, I want that.

Comment (sebasfalcone) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:215]: On it then!

Comment (sebasfalcone) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:213]: The problem with not waiting for the server to initialize correctly is that servers will start in a false state for a time even if they are in green (givin error/warn logs). I will see how I can circumvent this

**Comment by sebasfalcone** [comment]
Some tests are failing in GA but passing locally:
```
python3 build.py --target agent --readytoreview wazuh_modules/syscollector
 <wazuh_modules/syscollector>=============== Running RTR checks  ===============<wazuh_modules/syscollector> 
 <wazuh_modules/syscollector>=============== Running cppcheck    ===============<wazuh_modules/syscollector> 
 [Cppcheck: PASSED] 
 <wazuh_modules/syscollector>=============== Running AStyle      ===============<wazuh_modules/syscollector> 
 [Cleanfolder : PASSED] 
 [AStyle Check: PASSED] 
 [CleanAll: PASSED] 
 <agent>=============== Running Make Deps   ===============<agent>
 [MakeDeps: PASSED] 
 <agent>=============== Running Make project ==============<agent>
 [MakeTarget: PASSED] 
 <wazuh_modules/syscollector>=============== Running Tests       ===============<wazuh_modules/syscollector> 
 [syscollectorimp_unit_test: PASSED] 
 [sys_normalizer_unit_test: PASSED] 
 <wazuh_modules/syscollector>[All tests: PASSED]<wazuh_modules/syscollector> 
 ```

**Comment by sebasfalcone** [comment]
## Update
Found the root cause of the failing test. We were previously overriding the default headers of the http-request library with this
```C++
const std::unordered_set<std::string> headers {"Accept-Charset: utf-8"};
```

This prevented us from getting a json response from the indexer, which is easier to parse and work with.

I've aligned the code with this new response type

**Comment by sebasfalcone** [comment]
## Testing
Start with a cluster in yellow:
```json
{
  "cluster_name": "wazuh-cluster",
  "status": "yellow",
  "timed_out": false,
  "number_of_nodes": 1,
  "number_of_data_nodes": 1,
  "discovered_master": true,
  "discovered_cluster_manager": true,
  "active_primary_shards": 169,
  "active_shards": 169,
  "relocating_shards": 0,
  "initializing_shards": 0,
  "unassigned_shards": 2,
  "delayed_unassigned_shards": 0,
  "number_of_pending_tasks": 0,
  "number_of_in_flight_fetch": 0,
  "task_max_waiting_in_queue_millis": 0,
  "active_shards_percent_as_number": 98.83040935672514
}
```

And an empty index:
```json
{
  "took": 0,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 0,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  }
}
```

After restarting the manager. No events indexed:
```json
{
  "took": 0,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 0,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  }
}
```

We fix the cluster:
```json
{
  "cluster_name": "wazuh-cluster",
  "status": "green",
  "timed_out": false,
  "number_of_nodes": 1,
  "number_of_data_nodes": 1,
  "discovered_master": true,
  "discovered_cluster_manager": true,
  "active_primary_shards": 169,
  "active_shards": 169,
  "relocating_shards": 0,
  "initializing_shards": 0,
  "unassigned_shards": 0,
  "delayed_unassigned_shards": 0,
  "number_of_pending_tasks": 0,
  "number_of_in_flight_fetch": 0,
  "task_max_waiting_in_queue_millis": 0,
  "active_shards_percent_as_number": 100
}
```

Events are now indexed:
```json
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 60,
      "relation": "eq"
    },
```

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 10]
```diff
@@ -19,61 +19,122 @@
...
+            //         "max_task_wait_time": "-",
+            //         "active_shards_percent": "100.0%"
+            //     }
+            // ]
+
+            const auto data = nlohmann::json::parse(response, nullptr, false).at(0);
+
+            // Check if the server is green
+            if (!data.is_discarded() && data.contains(SERVER_HEALTH_FIELD_NAME) &&
+                data.at(SERVER_HEALTH_FIELD_NAME).get_ref<const std::string&>().compare("green") == 0)
```
[the remark this answers is not in the document]
Comment (Dwordcito) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:98]: Based on this documentation: https://www.elastic.co/guide/en/elasticsearch/reference/current/red-yellow-cluster-status.html

Difference Between Yellow and Red Status:

Yellow Status:
	•	Primary shards are allocated, meaning all the data is accessible.
	•	Some replica shards are unassigned, which reduces fault tolerance since there’s no replica of certain data on other nodes.
	•	The data is still readable and writable, but in the event of another node failure, data availability may be compromised due to the lack of replicas.
Red Status:
	•	One or more primary shards are unallocated, meaning some of the data is inaccessible and possibly lost.
	•	Data loss is a real risk in this state, and the cluster can’t fully serve read or write operations for the unallocated primary shards.
	•	The cluster is in a critical state, and immediate action is needed to recover or reallocate the primary shards to restore data availability.

In summary, yellow status means the cluster is operational but lacks fault tolerance due to missing replicas, while red status indicates that some data is unavailable and needs immediate attention.

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 11]
```diff
@@ -19,61 +19,122 @@
...
+constexpr auto SERVER_HEALTH_FIELD_NAME {"status"};
 
 /**
  * @brief Monitoring class.
  *
  */
 class Monitoring final
 {
-    std::map<std::string, bool> m_values;
+    std::map<std::string, bool> m_servers;
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:46]: Sonar lint was suggesting the same
image

File: src/shared_modules/indexer_connector/src/monitoring.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/shared_modules/indexer_connector/src/monitoring.hpp:98]: I will update the logic to allow the use of yellow clusters

File: src/shared_modules/indexer_connector/src/monitoring.hpp [hunk 12]
```diff
@@ -14,66 +14,127 @@
...
 #include <condition_variable>
+#include <json.hpp>
 #include <map>
 #include <mutex>
 #include <string>
 #include <thread>
-#include <unordered_set>
 #include <vector>
 
+constexpr auto MONITORING_NAME {"monitoring"};
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/monitoring.hpp:27]: unused
```suggestion
```

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 13]
```diff
@@ -261,14 +274,21 @@ void IndexerConnector::diff(const nlohmann::json& responseJson,
 
     if (!bulkData.empty())
     {
+        const auto onSuccess = [](const std::string& response)
+        {
+            logDebug2(IC_NAME, "Response: %s", response.c_str());
+        };
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:280]: unused

```suggestion

```

**Review by Dwordcito — COMMENTED** [review]
unused lambda

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp (hunk 13 above)
Comment (sebasfalcone) [reply src/shared_modules/indexer_connector/src/indexerConnector.cpp:280]: Instead I added the lambda to improve debugging

**Review by Dwordcito — APPROVED** [review]
